app/src/main/java/org/dpppt/android/app/contacts/ContactsFragment.java

```diff
@@ -16,113 +17,124 @@
-																: R.string.tracing_error_text;
-			TracingStatusHelper.updateStatusView(contactStatusView, state, titleRes, textRes);
-
-			invalidateErrorResolverButtons();
-		});
-
-		locationPermissionButton = view.findViewById(R.id.contact_location_permission_button);
-		locationPermissionButton.setOnClickListener(v -> {
-			requestedSomething = true;
-			String[] permissions = new String[] { Manifest.permission.ACCESS_FINE_LOCATION };
-			requestPermissions(permissions, 1);
-		});
-
-		batteryOptimizationButton = view.findViewById(R.id.contact_battery_button);
-		batteryOptimizationButton.setOnClickListener(v -> {
-			requestedSomething = true;
-			startActivity(new Intent(Settings.ACTION_REQUEST_IGNORE_BATTERY_OPTIMIZATIONS,
-					Uri.parse("package:" + requireContext().getPackageName())));
-		});
-
-		bluetoothButton = view.findViewById(R.id.contact_bluetooth_button);
-		bluetoothButton.setOnClickListener(v -> {
-			Toast.makeText(v.getContext(), getString(R.string.activate_bluetooth_button) + " ...", Toast.LENGTH_SHORT).show();
-			BluetoothAdapter.getDefaultAdapter().enable();
-		});
-		tracingViewModel.getBluetoothEnabledLiveData().observe(getViewLifecycleOwner(), bluetoothEnabled -> {
-			bluetoothButton.setVisibility(bluetoothEnabled ? View.GONE : View.VISIBLE);
-		});
-
-		invalidateErrorResolverButtons();
-	}
-
-	@Override
-	public void onResume() {
-		super.onResume();
-
-		if (requestedSomething) {
-			tracingViewModel.invalidateService();
-			invalidateErrorResolverButtons();
-			requestedSomething = false;
-		}
-	}
-
-	private void invalidateErrorResolverButtons() {
-		boolean locationPermissionGranted = DeviceFeatureHelper.isLocationPermissionGranted(requireContext());
-		locationPermissionButton.setVisibility(locationPermissionGranted ? View.GONE : View.VISIBLE);
-
-		boolean batteryOptDeactivated = DeviceFeatureHelper.isBatteryOptimizationDeactivated(requireContext());
-		batteryOptimizationButton.setVisibility(batteryOptDeactivated ? View.GONE : View.VISIBLE);
-	}
+    private TracingViewModel tracingViewModel;
+
+    private Button locationPermissionButton;
+    private Button bluetoothButton;
+    private Button batteryOptimizationButton;
+
+    private boolean requestedSomething = false;
+
+    @Inject
+    ViewModelFactory viewModelFactory;
+
+    public static ContactsFragment newInstance() {
+        return new ContactsFragment();
+    }
+
+    public ContactsFragment() {
+        super(R.layout.fragment_contacts);
+    }
+
+    @Override
+    public void onAttach(@NonNull Context context) {
+        super.onAttach(context);
+        MainApplication.Companion.getAppComponent(context).inject(this);
+    }
+
+    @Override
+    public void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        tracingViewModel = new ViewModelProvider(requireActivity(), viewModelFactory).get(TracingViewModel.class);
```

Comment:
same here try to wrap as much as you can inside utils methods in order to ensure that everyone can benefit your implementation.